File: UnitConvertor/app/src/main/java/xinyuz0416/android/unitconvertor/MainActivity.java
package xinyuz0416.android.unitconvertor;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    EditText editText;
    Button btn, btn2;
    TextView textView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        editText = findViewById(R.id.editText);
        btn = findViewById(R.id.btn);
        btn2 = findViewById(R.id.btn2);
        textView = findViewById(R.id.textView);

        btn.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                String input = editText.getText().toString();
                double kilos = Double.parseDouble(input);
                textView.setText(""+kilosToPounds(kilos));
            }
        });

        btn2.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                String input = editText.getText().toString();
                double pounds = Double.parseDouble(input);
                textView.setText(""+PoundsToKilos(pounds));
            }
        });


    }

    public double kilosToPounds(double kilos){
        return kilos*2.20462;
    }
    public double PoundsToKilos(double pounds){
        return pounds*0.45359;
    }
}

File: GeoQuiz/app/src/main/java/xinyuz0416/android/geoquiz/MainActivity.kt
package xinyuz0416.android.geoquiz

import android.app.Activity
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.*
import android.util.*
import android.view.*
import android.widget.*
import androidx.activity.*
import androidx.activity.result.contract.ActivityResultContracts
import xinyuz0416.android.geoquiz.databinding.ActivityMainBinding

private const val TAG = "MainActivity"

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    //    property delegate
    private val quizViewModel: QuizViewModel by viewModels()
    //    register MainActivity for an ActivityResult to hear back from CheatActivity to know if user cheated
    private val cheatLauncher = registerForActivityResult(
                                    ActivityResultContracts.StartActivityForResult()
                                ){result-> //handle the result
                                    if(result.resultCode == Activity.RESULT_OK){
                                        quizViewModel.isCheater = true
                                    }
                                }

    //    —————————————————————————————— onCreate —————————————————————————————————————
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG,"onCreate(Bundle?) called")
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        Log.d(TAG,"Got a QuizViewModel: $quizViewModel")

        updateQuestion()


        binding.trueButton.setOnClickListener {view: View ->
            checkAnswer(true)
        }
        binding.falseButton.setOnClickListener { view: View, ->
            checkAnswer(false)
        }
        binding.nextButton.setOnClickListener {
            quizViewModel.moveToNext()
            updateQuestion()
        }
        binding.prevButton.setOnClickListener {
            quizViewModel.moveToPrev()
            updateQuestion()
        }
        binding.cheatButton.setOnClickListener {
            // Start CheatActivity
            val answerIsTrue = quizViewModel.currentQuestionAnswer //get the current answer
            val intent = CheatActivity.newIntent(this@MainActivity,answerIsTrue) //create an intent while passing in current answer
            cheatLauncher.launch(intent) //start activity for result
        }


        binding.questionTextView.setOnClickListener {
            quizViewModel.moveToNext()
            updateQuestion()
        }

    }




    private fun updateQuestion(){
        val questionResId = quizViewModel.currentQuestionText
        binding.questionTextView.setText(questionResId)
    }

    private fun checkAnswer(userAnswer: Boolean){
        val correctAnswer = quizViewModel.currentQuestionAnswer

        val messageResId = when{
            quizViewModel.isCheater -> R.string.judgment_toast
            userAnswer == correctAnswer -> R.string.correct_toast
            else -> R.string.incorrect_toast
        }
        Toast.makeText(this,messageResId, Toast.LENGTH_SHORT).show()

    }

    override fun onStart() {
        super.onStart()
        Log.d(TAG, "onStart() called")
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume() called")
    }

    override fun onPause(){
        super.onPause()
        Log.d(TAG, "onPause() called")
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop() called")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy() called")
    }
}